### PR TITLE
Add HLLC numerical flux

### PR DESCRIPTION
**Pull request by mwarusz**
Co-authored-by: Thomas Gibson <thomas.gibson@nps.edu>

# Description

Adds HLLC numerical flux.

I have

- [ ] Written and run all necessary tests with CLIMA by including `tests/runtests.jl`
- [ ] Followed all necessary [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) and run `julia .dev/climaformat.jl .`
- [ ] Updated the documentation to reflect changes from this PR.

# For review by CLIMA developers

- [x] There are no open pull requests for this already
- [ ] CLIMA developers with relevant expertise have been assigned to review this submission
- [ ] The code conforms to the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) and has consistent naming conventions. `julia .dev/format.jl` has been run in a separate commit.
- [ ] This code does what it is technically intended to do (all numerics make sense physically and/or computationally)

### CODE REVIEW

**Comment by mwarusz** [comment]
bors try

File: src/Atmos/Model/AtmosModel.jl
Comment (charleskawczynski) [inline src/Atmos/Model/AtmosModel.jl:945]: Is this expected to work at the boundaries? At the boundary, `state_auxiliary⁻` (specifically variables that are a function of the state) is not necessarily in sync with the state assigned in `boundary_state!`, in which case, you'll get different results by creating a new thermodynamic state:
```julia
ts⁻ = PhaseEquil(
        atmos.param_set,
        internal_energy(atmos, state, aux),
        state⁻.ρ,
        state⁻.moisture.ρq_tot / state.ρ,
        moist.maxiter,
        atmos.moisture.tolerance,
    )
```
The problem with this is that it will call saturation adjustment.

File: src/Atmos/Model/AtmosModel.jl
Comment (charleskawczynski) [inline src/Atmos/Model/AtmosModel.jl:967]: Is this expected to work at the boundaries? At the boundary, `state_auxiliary⁺` (specifically variables that are a function of the state) is not necessarily in sync with the state assigned in `boundary_state!`, in which case, you'll get different results by creating a new thermodynamic state:
```julia
ts⁺ = PhaseEquil(
        atmos.param_set,
        internal_energy(atmos, state, aux),
        state⁺.ρ,
        state⁺.moisture.ρq_tot / state.ρ,
        moist.maxiter,
        atmos.moisture.tolerance,
    )
```
The problem with this is that it will call saturation adjustment.

Comment (jkozdon) [reply src/Atmos/Model/AtmosModel.jl:967]: Shouldn't this always be the case though? If you modify the state in a way that requires a saturation adjustment.

Comment (charleskawczynski) [reply src/Atmos/Model/AtmosModel.jl:967]: There are two ways to compute thermo states. One that calls SA (ideally once in `nodal_update_auxiliary_state!`)
```julia
ts = PhaseEquil(
        atmos.param_set,
        internal_energy(atmos, state, aux),
        state.ρ,
        state.moisture.ρq_tot / state.ρ,
        moist.maxiter,
        atmos.moisture.tolerance,
    )
```
And one that _recovers_ the thermodynamic state:
```julia
ts = thermo_state(atmos, state, aux)
```
This second method grabs the temperature from the `aux` state, and calls the `PhaseEquil{...}` (the default constructor) _with_ temperature (computed in `nodal_update_auxiliary_state!`), so SA is never performed with this second method.

The idea was that users should only need to call the second method. This assumes that the `aux` and `prognostic` states are in sync, but this may be invalid if we're updating/changing the prognostic state (as is the case on the boundaries).

Comment (mwarusz) [reply src/Atmos/Model/AtmosModel.jl:967]: Sounds like there is no way to avoid SA if we need pressure that is consistent with the boundary state ? Wondering where it should be performed. Other numerical fluxes don't need it so doing it  in `boundary_state!` would be wasteful. Maybe we should define a special method of `numerical_boundary_flux_first_order!` for `HLLC` and `Roe`. Or is it better to do it outside of surface kernels and communicate through aux like for volume SA ?

**Comment by jkozdon** [comment]
> Or is it better to do it outside of surface kernels and communicate through aux like for volume SA ?

Not sure this would be possible since we do not have aux storage for the plus side of the boundary (the plus side is set to the minus side on the boundary).

This could change, but we'd have to modify the mesh generation and array creation.

File: src/Atmos/Model/AtmosModel.jl
Comment (jkozdon) [reply src/Atmos/Model/AtmosModel.jl:967]: > Or is it better to do it outside of surface kernels and communicate through aux like for volume SA ?

Not sure this would be possible since we do not have aux storage for the plus side of the boundary (the plus side is set to the minus side on the boundary).

This could change, but we'd have to modify the mesh generation and array creation.

Comment (charleskawczynski) [reply src/Atmos/Model/AtmosModel.jl:967]: I'm not sure what the best way forward is, but it seems like, for now, do the correct thing and call SA. We should probably have a discussion regarding this (I imagine that this may be problematic in several places).

Comment (charleskawczynski) [reply src/Atmos/Model/AtmosModel.jl:998]: Maybe we can meet to discuss this more?

Comment (thomasgibson) [reply src/Atmos/Model/AtmosModel.jl:998]: Agreed. I think we should definitely talk about this

Comment (jkozdon) [reply src/Atmos/Model/AtmosModel.jl:998]: It's possible that we can get away with an SA in the boundary state. We won't know until the code is profiled to see if we had register spilling or other problems.

We moved it to aux initially due to performance issues, but this situation might not be as bad.

Also, is it clear that we shouldn't be doing it for all numerical fluxes? (All the fluxes evaluate the flux with the boundary state which in turn involves pressure and wavespeed)

Comment (mwarusz) [reply src/Atmos/Model/AtmosModel.jl:998]: You're right. All fluxes have the plus state pressure and sound speed inconsistent with BCs when using `EquilMoist`. I will check if performing SA in `boundary_state!` is a performance problem.

Comment (charleskawczynski) [reply src/Atmos/Model/AtmosModel.jl:998]: > It's possible that we can get away with an SA in the boundary state. 

Agreed. Let's try it.

> We won't know until the code is profiled to see if we had register spilling or other problems.

✔️ 

> Agreed. I think we should definitely talk about this

While we have a quick/easy first pass at a solution to this particular problem, I agree that we should still talk about the general concern of accidentally using `aux⁺` (in the case that it's not in sync with `state⁺`).

File: src/Atmos/Model/AtmosModel.jl
Comment (charleskawczynski) [inline src/Atmos/Model/AtmosModel.jl:976]:
```suggestion
    ts⁻ = recover_thermo_state(
```

File: src/Atmos/Model/AtmosModel.jl
Comment (charleskawczynski) [inline src/Atmos/Model/AtmosModel.jl:998]:
```suggestion
    ts⁺ = new_thermo_state(
```

File: src/Atmos/Model/AtmosModel.jl
Comment (charleskawczynski) [inline src/Atmos/Model/AtmosModel.jl:1013]:
```suggestion
    p⁺ = air_pressure(ts⁺)
```

Comment (charleskawczynski) [reply src/Atmos/Model/AtmosModel.jl:1013]: I'm now thinking that we should remove `pressure`, since this relies on `recover_thermo_state` and will have the same issue as above. Using `ts⁺ = new_thermo_state(...)` will fix these issues.

File: src/Atmos/Model/AtmosModel.jl
Comment (charleskawczynski) [inline src/Atmos/Model/AtmosModel.jl:993]:
```suggestion
    p⁻ = air_pressure(ts⁻)
```
This one doesn't matter as much, but we might as call `air_pressure` directly.

File: src/Atmos/Model/AtmosModel.jl
Comment (mwarusz) [reply src/Atmos/Model/AtmosModel.jl:998]: Should this become `recover_thermo_state` after #1542 gets merged ?

Comment (charleskawczynski) [reply src/Atmos/Model/AtmosModel.jl:998]: Ah, yes. So long `numerical_flux_first_order!` is called right after `boundary_state!`.

Comment (mwarusz) [reply src/Atmos/Model/AtmosModel.jl:998]: It is https://github.com/CliMA/ClimateMachine.jl/blob/721b67f2fc72692fe4daf63d9c091d3398b1b889/src/Numerics/DGMethods/NumericalFluxes.jl#L147-L171

**Comment by bischtob** [comment]
@thomasgibson, @mwarusz is this good to go pending @charleskawczynski's changes?

File: src/Atmos/Model/AtmosModel.jl
Comment (mwarusz) [reply src/Atmos/Model/AtmosModel.jl]: Done

File: src/Atmos/Model/AtmosModel.jl
Comment (mwarusz) [reply src/Atmos/Model/AtmosModel.jl]: Done

File: src/Atmos/Model/AtmosModel.jl
Comment (mwarusz) [reply src/Atmos/Model/AtmosModel.jl]: Done

File: src/Atmos/Model/AtmosModel.jl
Comment (mwarusz) [reply src/Atmos/Model/AtmosModel.jl]: Done

**Comment by mwarusz** [comment]
bors try

**Review by simonbyrne — APPROVED** [review]

**Pull request merged**
